### PR TITLE
Auto is the default interface

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**

`auto` interface is currently supported but not default, this PR makes it default.

**Description of the Change:**

1. It makes all transforms compatible with `auto`: interface is defined in `__call__` instead of `__init__` of the QNode as well as gradient function and gradient kwargs, also potentially the device is switched in call instead of init.
2. It updates all transforms tests to add a `auto` interface
3. As the QNode are tested in the interfaces, the interfaces are removed for other tests.

**Benefits:**

No need to define interface when you give parameters.

**Possible Drawbacks:**

1. Small overhead to get the interface from the args.
2. Unable to track a device that is changing in `__call__` instead of `__init__` if you use

**Example**

```
  import jax
  import jax.numpy as jnp
  
  qml.enable_return()
  a = jnp.array(0.1)
  b = jnp.array(0.2)
  
  dev = qml.device("default.qubit", wires=2)
  
  
  @qml.qnode(dev)
  def circuit(a, b):
      qml.RY(a, wires=0)
      qml.RX(b, wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))
  ```
  ```pycon
  >>> circuit(a, b)
  (Array(0.9950042, dtype=float32), Array(-0.19767681, dtype=float32))
  
  >>> jac = jax.jacobian(circuit)(a, b)
  (Array(-0.09983341, dtype=float32, weak_type=True), Array(0.01983384, dtype=float32, weak_type=True))
```

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-31368]

File: pennylane/devices/tests/test_measurements.py [hunk 1]
```diff
@@ -1615,9 +1615,10 @@ def circuit():
     def test_method_overriden_by_device(self, device):
         """Test that the device can override a measurement process."""
         dev = device(2)
+
         _skip_test_for_braket(dev)
 
-        @qml.qnode(dev)
+        @qml.qnode(dev, interface="autograd")
```
Comment (rmoyard) [inline pennylane/devices/tests/test_measurements.py:1621]: Here defining the interface is needed because `update_gradient_fn` in `__init__` is updating the device, as with `auto` this step is kipped in `__init__`. Then we apply the changes on the device before it is transformed and therefore loose the custom method.

File: pennylane/debugging.py [hunk 2]
```diff
@@ -81,6 +82,9 @@ def circuit():
     """
 
     def get_snapshots(*args, **kwargs):
+        if qnode.interface == "auto":
+            qnode.interface = qml.math.get_interface(*args, *list(kwargs.values()))
```
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/debugging.py:86]: > Maybe use a with statement?

We could add a `set_interface` contextmanager method to the `QNode`:
```python
@contextlib.contextmanager
def set_interface(self, *args, **kwargs):
    if self.interface != "auto":
        yield
        return
    self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
    yield
    self.interface = "auto"
    return
```
Such that we can do:
```python
with qnode.set_interface(*args, **kwargs):
    # do sth
```

[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/debugging.py:86]: Here's another potential solution to this problem:

I believe this interface problem comes when wrapping `QNodes`, right? Thus all wrappers will have the same signature: `def wrapper(*args, **kwargs)`. We could then define the following method inside `QNode`:

```python
def set_interface(self, f):
    old_interface = self.interface
    def inner(*args, **kwargs):
      if self.interface == "auto":
          self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
      result = f(*args, **kwargs)
      self.interface = old_interface
      return result
  return inner
```
Such that we can do:
```python
def transform(tape):
    def wrapper(*args, **kwargs):
        # do sth
    return tape.set_interface(wrapper)  # we should find a better name
```

**PRO**: No need of a context manager. Avoid indenting long functions.
**CON**: `f` must have the `f(*args, **kwargs)` signature.

**Review by AlbertMitjans — COMMENTED** [review]
Amazing job!! 💯 🚀  This PR must've been difficult to write...

My main concern is that we should set `qnode.interface` back to auto in every transform. I left some suggestions to solve this problem.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -388,6 +388,41 @@
 
 <h3>Breaking changes</h3>
 
+* The default interface is now `auto`, there is no need to specify the interface anymore! It is automatically
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:391]:
```suggestion
* The default interface is now `auto`. There is no need to specify the interface anymore! It is automatically 
```
Or potentially using a semicolon if you really don't want to use a period.

File: tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py [hunk 4]
```diff
@@ -347,7 +349,7 @@ def test_hessian_expval_multiple_params(
     ):
         """The hessian of a single measurement with multiple params return a tuple of arrays."""
 
-        if interface == "tf" and diff_method == "spsa":
+        if (interface == "tf" or interface == "auto") and diff_method == "spsa":
             # TODO: Find out why.
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py:353]: No I don't think it is tracked, @dwierichs maybe you know more about that.

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -388,6 +388,41 @@
...
+  >>> circuit(a, b)
+  (Array(0.9950042, dtype=float32), Array(-0.19767681, dtype=float32))
+  >>> jac = jax.jacobian(circuit)(a, b)
+  (Array(-0.09983341, dtype=float32, weak_type=True), Array(0.01983384, dtype=float32, weak_type=True))
+  ```
+  
+  It comes with a small overhead and the fact that the interface is determined during the `QNode` call instead of the 
+  initialization. It means that the `gradient_fn` and `gradient_kwargs` are only defined on the QNode at the beginning 
+  of the call. Also without specifying the interface it is not possible to guarantee that the device will not be changed 
+  during the call if you are using backprop(`default.qubit` to `default.qubit,jax`e.g.) where before it was happening at
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply doc/releases/changelog-dev.md:435]: Here we speak about the device not the interface, because the interface control the device in `_update_gradient_fn`

File: doc/introduction/interfaces/jax.rst [hunk 6]
```diff
@@ -75,6 +75,9 @@ as input, and returning ``jax.Array`` objects. It can now be used like any other
 >>> circuit1(phi, theta)
 Array([0.8776, 0.6880], dtype=float64)
 
+The interface can also be automatically determined when the ``QNode`` is called. You do not need to pass the interface
+if you provide parameters, but it comes with a small overhead.
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply doc/introduction/interfaces/jax.rst:79]: I removed it :+1:

File: pennylane/devices/tests/test_measurements.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/devices/tests/test_measurements.py:1620]: If you use backprop you need the interface, because `_update_gradient_fn` will potentially change the device. I think it will be easier to adress after the device revamp.

File: pennylane/ops/functions/map_wires.py [hunk 7]
```diff
@@ -117,6 +117,8 @@ def qfunc(*args, **kwargs):
             return m[0] if len(m) == 1 else m
 
         if isinstance(input, QNode):
+            if input.gradient_kwargs is None:
+                input.gradient_kwargs = {}
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/ops/functions/map_wires.py:120]: Yes I will change that :+1:

File: pennylane/devices/tests/test_measurements.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/devices/tests/test_measurements.py:1620]: that makes sense, I'm just worried that someone will not specify this because they didn't think they needed to? idk

Comment (rmoyard) [reply pennylane/devices/tests/test_measurements.py:1620]: Yes I agree it is not ideal, we can add a warning if the device is changed during `__call__`. It seems to be a good compromise?

File: tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py:353]: No, sorry. That's on me, I missed to track it.

[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py:353]: #3715

[the remark this answers is not in the document]
Comment (timmysilv) [reply tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py:353]: thanks!

File: pennylane/devices/tests/test_measurements.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/devices/tests/test_measurements.py:1620]: tbh I don't think a user would hit it anyway... we can cross that bug-bridge when we get there. Largely because I'm not sure how I'd warn. save a hash of the measurement map (or the whole device?) at the end of qnode init, validate that the hash is the same at the beginning of qnode call?

Comment (rmoyard) [reply pennylane/devices/tests/test_measurements.py:1620]: I was speaking not specifically for this test but more in general raise a warning if the device is switched in the `__call__`

Comment (timmysilv) [reply pennylane/devices/tests/test_measurements.py:1620]: I see. I think I was looking for something a bit more general, but you're proposing to raise a warning [here](https://github.com/PennyLaneAI/pennylane/blob/876bac85c1b84c00368b7e7d34ecfcbd8b9c59b3/pennylane/qnode.py#L653-L659)?

File: pennylane/debugging.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/debugging.py]: Great catch thank you :100: I did it less automatically because sometimes you call a transform in the transform and do not execute the QNode. I have added a couple of assert in transforms test.

File: pennylane/devices/tests/test_measurements.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/devices/tests/test_measurements.py:1620]: I added the warning and a test (test_qnode.py) let me know what you think!

File: pennylane/qnode.py [hunk 8]
```diff
@@ -728,6 +728,9 @@ def tape(self) -> QuantumTape:
     def construct(self, args, kwargs):
         """Call the quantum function with a tape context, ensuring the operations get queued."""
 
+        if self.interface == "auto":
+            self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
```
Comment (AlbertMitjans) [inline pennylane/qnode.py:732]: Need to set interface back to "auto"

File: pennylane/qnode.py [hunk 9]
```diff
@@ -793,8 +796,14 @@ def construct(self, args, kwargs):
...
         override_shots = False
         old_interface = self.interface
+
         if old_interface == "auto":
             self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
+            if self.device is not self._original_device:
+                warnings.warn(
+                    "The device was switched during the call of the QNode, to avoid this behaviour define"
+                    "an interface argument instead of auto."
+                )
```
Comment (AlbertMitjans) [inline pennylane/qnode.py:806]: 💯

File: tests/measurements/test_mutual_info.py [hunk 10]
```diff
@@ -206,7 +207,7 @@ def test_qnode_mutual_info_jax_jit(self, params):
 
         params = jnp.array(params)
 
-        @qml.qnode(dev, interface="jax-jit")
+        @qml.qnode(dev, interface=interface)
```
Comment (AlbertMitjans) [inline tests/measurements/test_mutual_info.py:210]: Why this change? It should have the same behaviour, right?

**Review by AlbertMitjans — APPROVED** [review]
Amazing job 💯 🚀 

I believe `QNode.construct` should set the interface back to "auto". Otherwise this PR is ready to be merged 🔥

File: tests/measurements/test_mutual_info.py (hunk 10 above)
Comment (rmoyard) [reply tests/measurements/test_mutual_info.py:210]: Yes my first attempt was to add auto everywhere but the test suite time double, so I just removed the auto interface but not the fixture in case we want to readd it at some point.

File: pennylane/qnode.py (hunk 8 above)
Comment (rmoyard) [reply pennylane/qnode.py:732]: I don't think so, we cannot pass it to execute if we reset in the construct and therefore we would need to recheck the interface in `execute` (currently it is reset at the end of the QNode call). Do you have a use case in mind?

File: tests/measurements/test_mutual_info.py (hunk 10 above)
Comment (josh146) [reply tests/measurements/test_mutual_info.py:210]: @rmoyard do you know why the test suite time doubled? Was it because of the `get_gradient_fn` being now done on call rather than on init?

Comment (rmoyard) [reply tests/measurements/test_mutual_info.py:210]: No it is because I added fixture for the interface everywhere, I would use ["auto", "jax", "jax-jit"] instead of "jax-jit" for example and therefore you triple the time for this specific test.

Comment (josh146) [reply tests/measurements/test_mutual_info.py:210]: ohhhh got it 😆 Thanks @rmoyard, this had me worried for a second

File: pennylane/qnode.py (hunk 8 above)
Comment (AlbertMitjans) [reply pennylane/qnode.py:732]: When calling a `QNode`, the `__call__` method switches from "auto" to the correct interface, and then switches back to "auto". But what if a user calls `construct` directly? Then we would lose the "auto"interface.

Comment (rmoyard) [reply pennylane/qnode.py]: Thanks done :+1:

**Comment by rmoyard** [comment]
auto_vs_interface
@AlbertMitjans @trbromley

**Comment by AlbertMitjans** [comment]
> auto_vs_interface @AlbertMitjans @trbromley For a small circuit 2 qubits circuit with 3 gates

Nice! 😍 

In which scenario will we execute multiple times the same circuit? For these cases we might use a wrapper that detects the interface only once, runs all the needed executions, and sets back the interface to "auto".
